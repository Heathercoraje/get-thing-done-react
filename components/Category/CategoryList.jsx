import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';


const Wrapper = styled.div`
	padding-left: 0.6rem;
	color: white;
`;
const CategoryList = ({ categories, ...props }) => (
	<Wrapper className="category-list">
		{categories.map(category => (
    <CategoryItem key={category.id} {...props} {...category} />
		))}
  </Wrapper>
);

CategoryList.propTypes = {
	props: PropTypes.object,
	categories: PropTypes.array,
	category: PropTypes.object,
	id: PropTypes.number
}
const CategoryItem = props => (
	<div className="category-div">
		<li
			onClick={props.click}
			value={props.id}
			data-name={props.name}
			className="category-text"
		>
			{props.name}
  </li>
		<button className="button-delete" onClick={props.delete} value={props.id}>
			&#8854;
  </button>
  </div>
);
CategoryItem.propTypes = {
	id: PropTypes.number,
	name: PropTypes.string,
	click: PropTypes.func,
	delete: PropTypes.func
}

export { CategoryList, CategoryItem };
