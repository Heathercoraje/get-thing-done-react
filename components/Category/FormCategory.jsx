import React, { Component } from 'react';
import HandleClickInput from '../../lib/HandleClickInput';

const FormCategory = ({ add }) => {
	const onClickInput = evt => {
		evt.preventDefault();
		const inputElement = document.getElementById('input-cate');
		HandleClickInput(evt, add, inputElement);
	};
	return (
		<form className="form-cate" onSubmit={onClickInput}>
			<input
				placeholder=" New category"
				id="input-cate"
				style={{ display: 'none' }}
				className="input-gen input-cate"
			/>
			<input type="submit" value="+" className="button-add" />
		</form>
	);
};

export default FormCategory;
