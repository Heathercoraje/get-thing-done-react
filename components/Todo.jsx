import React, { Component } from 'react';
import styled from 'styled-components';
import TodoList from './TodoList';
import FormTodo from './FormTodo';

const Wrapper = styled.div`
	box-sizing: border-box;
	border-top-right-radius: 0.2rem;
	border-bottom-right-radius: 0.2rem;
	background: white;
	padding: 3rem 1.5rem 1.5rem 2.5rem;
	float: right;
	width: 65%;
	height: 100%;
	display: flex;
	flex-direction: column;
	color: #393939;
`;

const HandleDesc = (e) => {
  if (e.keyCode === 13) {
    document.getElementById('descForm').submit();
  }
};

class Todo extends Component {
  constructor(props) {
    super(props);
    this.state = {
      description: 'hello world'
    };
    this.HandleDesc = this.HandleDesc.bind(this);
  }
  HandleDesc(e) {
    if (e.keyCode === 13) {
      this.setState({
        description: e.target.value
      });
      e.preventDefault();
    }
  }
  render() {
    return (
      <Wrapper>
        <TodoList
          categories={this.props.categories}
          todos={this.props.todos}
          selected={this.props.selected}
          delete={this.props.delete}
          HandleDesc={this.HandleDesc}
          desc={this.state.description}
        />
        <FormTodo add={this.props.add} selected={this.props.selected} />
      </Wrapper>
    );
  }
}

export default Todo;
