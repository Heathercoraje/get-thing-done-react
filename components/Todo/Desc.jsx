import React, { Component } from 'react';
import ReactDOM from 'react';
import styled from 'styled-components';

const DescBox = styled.div`
	margin-bottom: 3vh;
`;

class Desc extends Component {
	render() {
		const name = this.props.name;
		const description = this.props.description;
		return (
			<DescBox className="desc-box">
				<DescForm
    name={name}
					description={description}
    HandleDesc={this.props.HandleDesc}
  />
  </DescBox>
		);
	}
}

class DescForm extends Component {
	state = {
		edit: false
	};
	componentWillReceiveProps() {
		this.setState({
			edit: false
		});
	}

	toggle = () => {
		const prev = this.state.edit;
		this.setState({
			edit: !prev
		});
	};

	render() {
		const name = this.props.name;
		const description = this.props.description;
		if (this.state.edit) {
			return (
				<div>
					<Header name={name} toggle={this.toggle} />
					<form
						className="form-desc"
						onSubmit={evt => {
							this.props.HandleDesc(evt);
							this.toggle();
						}}
					>
						<input
							size="22"
							className="input-desc"
							type="text"
							placeholder={description}
							onBlur={evt => this.props.HandleDesc(evt)}
							autoFocus
						/>
						<input style={{ display: 'none' }} type="submit" />
					</form>
				</div>
			);
		} else {
			return (
				<div>
					<Header name={name} toggle={this.toggle} />
					<div id="desc" onClick={this.toggle}>
						{description}
					</div>
				</div>
			);
		}
	}
}

const Header = props => {
	return (
		<div className="cate-title">
			<h1 className="todo-name">{props.name}</h1>
			<button type="button" className="button-edit" onClick={props.toggle}>
				&nbsp;&#10000;
			</button>
		</div>
	);
};

export { Desc, DescForm, Header };
