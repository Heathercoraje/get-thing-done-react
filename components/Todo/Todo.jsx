import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { TodoList } from './TodoList';
import FormTodo from './FormTodo';
import localStorageApi from '../../lib/localStorageApi';


const Wrapper = styled.div`
  padding: 3rem 1.5rem 1.5rem 2.5rem;
	background: white;
	display: flex;
	flex-direction: column;
	color: #393939;
`;
class Todo extends Component {
	state = {
		todos: {}
	};

  componentWillMount () {
    localStorageApi.loadTodos().then((todos)=>{
      console.log('Fetching your data, friend.');
      this.setState({todos});
    });
  }

	componentWillReceiveProps(Nextprops) {
		this.updateTodoObj(Nextprops);
	}

	updateTodoObj = Nextprops => {
		const todoObj = Object.assign({}, this.state.todos);
		const updatedCategories = Nextprops.categories.map(
			category => category.name
		);
		for (const name in todoObj) {
			if (updatedCategories.indexOf(name) < 0) {
				delete todoObj[name];
			}
			continue;
		}
    localStorageApi.modifyTodos(todoObj);
		this.setState({
			todos: todoObj
		});
	};
	addTodo = evt => {
		const allTodos = this.state.todos;
		const selected = this.props.selected;
		const category = this.props.categories[selected];
		const name = category.name;
		// if this is a new categoty add new key of category and add empty array
		if (!allTodos[`${name}`]) {
			allTodos[`${name}`] = [];
		}
		const Todo = allTodos[`${name}`];
		const newTodo = {
			id: Todo.length === 0 ? 0 : Todo[Todo.length - 1].id + 1,
			todo: evt.target[0].value,
			isDone: false
		};
		Todo.push(newTodo);
    localStorageApi.modifyTodos(allTodos).then(()=>{
      this.setState({
        todos: allTodos
      });
      console.log('todo item added');
    });
	};
	deleteTodo = evt => {
		const allTodos = this.state.todos;
		const selected = this.props.selected;
		const category = this.props.categories[selected];
		const name = category.name;
		const newTodos = allTodos[`${name}`].filter(
			todo => todo.id !== Number(evt.target.value)
		);
		allTodos[`${name}`] = newTodos;
    localStorageApi.modifyTodos(allTodos).then(()=>{
      this.setState({
        todos: allTodos
      });
      console.log('todo item deleted');
    });

	};
	completeTodo = evt => {
		const allTodos = this.state.todos;
		const selected = this.props.selected;
		const category = this.props.categories[selected];
		const name = category.name;
		const targetTodoArr = allTodos[`${name}`];
		const targetId = evt.target.id;
		targetTodoArr.forEach(todo => {
			if (todo.id == targetId) {
				let temp = todo.isDone;
				todo.isDone = !temp;
			}
		});
		allTodos[`${name}`] = targetTodoArr;
    localStorageApi.modifyTodos(allTodos).then(()=>{
      this.setState({
        todos: allTodos
      });
      console.log('todo item completed');
    });

	};

	render() {
		const selected = this.props.selected;
		if (selected < 0) {
			return (
				<Wrapper>
					<h3> Add a new category </h3>{' '}
				</Wrapper>
			);
		}
		const category = this.props.categories[selected];
		const name = category.name;
		return (
			<Wrapper className="todo-container">
				<TodoList
					todos={this.state.todos[`${name}`]}
					selected={selected}
					name={name}
					categories={this.props.categories}
					HandleDesc={this.props.HandleDesc}
					delete={this.deleteTodo}
					complete={this.completeTodo}
    />{' '}
				<FormTodo add={this.addTodo} />{' '}
			</Wrapper>
		);
	}
}

Todo.propTypes = {
  categories: PropTypes.array,
  selected: PropTypes.number,
  HandleDesc: PropTypes.func,
}

export default Todo;
