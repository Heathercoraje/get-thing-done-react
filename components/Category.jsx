import React, { Component } from 'react';
import styled from 'styled-components';
import CategoryList from './CategoryList';
import FormCategory from './FormCategory';

const Wrapper = styled.div`
	box-sizing: border-box;
	border-top-left-radius: 10px;
	border-bottom-left-radius: 10px;
	padding: 20px;
	background: #393939;
	float: left;
	width: 35%;
	height: 100%;
	display: flex;
	flex-direction: column;
`;

const Title = <h2 className="Title">Get Things Done</h2>;

const Category = props => (
  <Wrapper>
    {Title}
    <CategoryList
      className="category-list"
      click={props.click}
      categories={props.categories}
      selected={props.selected}
      delete={props.delete}
    />
    <FormCategory add={props.add} />
  </Wrapper>
);

export default Category;
