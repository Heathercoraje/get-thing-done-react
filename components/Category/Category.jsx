import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { CategoryList } from './CategoryList';
import FormCategory from './FormCategory';

const Wrapper = styled.div`
	background: #393939;
	border-top-left-radius: 0.2rem;
	border-bottom-left-radius: 0.2rem;
	padding: 2.8rem 1rem 1.5rem 1.5rem;
	box-sizing: border-box;
	display: flex;
	flex-direction: column;
`;

const Category = props => (
	<Wrapper className="category-container">
		<h3 className="logo">Get Things Done</h3>
		<CategoryList {...props} />
		<FormCategory add={props.add} />
  </Wrapper>
);

Category.propTypes = {
	props: PropTypes.object,
	add: PropTypes.func
}

export default Category;
