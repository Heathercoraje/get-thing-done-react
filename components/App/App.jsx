import React, { Component } from 'react';
import styled from 'styled-components';
import Category from '../Category/Category';
import Todo from '../Todo/Todo';
import CheckDupli from '../../lib/CheckDupli';

const Wrapper = styled.div`
	margin: auto auto;
	display: flex;
`;

class App extends Component {
	state = {
		categories: [
			{ id: 0, name: 'Todo', description: 'Add description of category' }
		],
		selectedCategory: 0
	};

	addCategory = event => {
		const Categories = this.state.categories;
		const Todos = this.state.todos;
		const Name =
			event.target[0].value.charAt(0).toUpperCase() +
			event.target[0].value.slice(1);
		if (CheckDupli(Name, Categories)) {
			alert('duplicate cannot be added!');
			return;
		}
		const newCategory = {
			id: Categories.length ? Categories[Categories.length - 1].id + 1 : 0,
			name: Name,
			description: 'New category'
		};
		Categories.push(newCategory);
		const toDisplay = Categories.length - 1;
		this.setState({
			categories: Categories,
			selectedCategory: toDisplay,
			todos: Todos
		});
	};
	deleteCategory = evt => {
		if (Number(evt.target.value) === 0) return;
		const Categories = this.state.categories;
		const PrevToDisplay = this.state.selectedCategory;
		const Todos = this.state.todos;
		const newCategories = Categories.filter(
			category => category.id !== Number(evt.target.value)
		);
		const toDisplay =
			Number(evt.target.value) === PrevToDisplay
				? PrevToDisplay - 1
				: newCategories.length - 1;
		this.setState({
			categories: newCategories,
			selectedCategory: toDisplay
		});
	};
	HandleDesc = evt => {
		evt.preventDefault();
		const { categories, selectedCategory } = this.state;
		const updatedCategory = {
			...categories[selectedCategory],
			description: !evt.target[0] ? evt.target.value : evt.target[0].value
		};

		evt.target.value = '';
		this.setState({
			categories: [
				...categories.slice(0, selectedCategory),
				updatedCategory,
				...categories.slice(selectedCategory + 1)
			]
		});
	};
	clickCate = evt => {
		const name = evt.target.dataset.name;
		let toDisplay;
		const test = this.state.categories.map((category, index) => {
			if (category.name === name) {
				toDisplay = index;
			}
		});
		this.setState({
			selectedCategory: toDisplay
		});
	};

	render() {
		return (
			<Wrapper className="main-container">
				<Category
					click={this.clickCate}
					categories={this.state.categories}
					selected={this.state.selectedCategory}
					delete={this.deleteCategory}
					add={this.addCategory}
				/>
				<Todo
					categories={this.state.categories}
					selected={this.state.selectedCategory}
					HandleDesc={this.HandleDesc}
				/>
			</Wrapper>
		);
	}
}

export default App;
