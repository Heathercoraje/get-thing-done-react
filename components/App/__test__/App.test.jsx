import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import App from '../App.jsx';
import Category from '../../Category/Category';
import Todo from '../../Todo/Todo';

test('App should render correrly', () => {
	const component = shallow(<App />);
	const tree = toJson(component);
	expect(tree).toMatchSnapshot();
});

test('App has Category and Todo components', () => {
	const wrapper = mount(<App />);
	expect(
		wrapper.containsAllMatchingElements([
			<div>
				<Category />
				<Todo />
			</div>
		])
	).toBe(true);
});
