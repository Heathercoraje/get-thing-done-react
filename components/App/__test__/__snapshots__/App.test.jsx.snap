// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correrly 1`] = `
<styled.div
  className="main-container"
>
  <Category
    add={[Function]}
    categories={
      Array [
        Object {
          "description": "Add description of category",
          "id": 0,
          "name": "Todo",
        },
      ]
    }
    click={[Function]}
    delete={[Function]}
    selected={0}
  />
  <Todo
    HandleDesc={[Function]}
    categories={
      Array [
        Object {
          "description": "Add description of category",
          "id": 0,
          "name": "Todo",
        },
      ]
    }
    selected={0}
  />
</styled.div>
`;
